Add initial draft of @wrap.check decorator

kaidokert opened the pull request:
Proposed way of doing  #282

hgrecco commented on docs/wrapping.rst:
```diff
@@ -147,5 +147,24 @@ To avoid the conversion of an argument or return value, use None
+
+.. doctest::
+
+    >>>@ureg.wraps('[length]')
```
this should be `check`

kaidokert commented:
Whoops i failed python3 really bad here. Will go back and fix

hgrecco commented on pint/unit.py:
```diff
@@ -1145,6 +1145,36 @@ def wrapper(*values, **kw):
+
+            @functools.wraps(func, assigned=assigned, updated=updated)
+            def wrapper(*values, **kwargs):
+                for dim, value in itertools.izip_longest(dimensions, values):
```
Two things about this.
1) izip_longest does not exist in Python 3+ (it is called zip_longest). Please add it via the compat module
2) In `wraps` we used zip, but actually I think that zip_longest is better. Would you mind submitting another PR fixing that.

hgrecco commented on pint/unit.py:
```diff
@@ -1145,6 +1145,36 @@ def wrapper(*values, **kw):
+            def wrapper(*values, **kwargs):
+                for dim, value in itertools.izip_longest(dimensions, values):
+                    if dim and value.dimensionality != dim:
+                        raise TypeError(
```
I think that a `DimensionlityError` would be a better choice

kaidokert replied:
Yeah i thought about this too, wasnt sure.

kaidokert commented:
Addressed most of the comments. I'll PR use of zip_longest as well after this. 
DimensionalityError that is thrown looks kinda funny now..
DimensionalityError: Cannot convert from '12 parsec' ([length]) to 'a quantity of' ([length] / [time])
But i think that would be mostly OK and understandable.

hgrecco commented:
This is merged but somehow GitHub is not recognizing it.
